pkg/serviceaccounttoken/secret.go

```diff
@@ -204,29 +205,84 @@ func isSecretForServiceAccount(secret *corev1.Secret, sa *corev1.ServiceAccount)
 	return sa.Name == annotation
 }
 
-func createServiceAccountSecret(ctx context.Context, sa *corev1.ServiceAccount, secretLister secretLister, secretClient clientv1.SecretInterface, lockKey string) (*corev1.Secret, error) {
-	mutex := getLock(lockKey)
-	mutex.Lock()
-	defer func(key string) {
-		mutex.Unlock()
-		lockMap.Delete(lockKey)
-	}(lockKey)
-
-	// We could have been waiting for the Mutex to unlock in a parallel run of
-	// createServiceAccountSecret - check again for the secret existing.
-	secret, err := ServiceAccountSecret(ctx, sa, secretLister, secretClient)
+func createServiceAccountSecret(ctx context.Context, sa *corev1.ServiceAccount, secretLister secretLister, secretClient clientv1.SecretInterface) (*corev1.Secret, error) {
+	sc := SecretTemplate(sa)
+	secret, err := secretClient.Create(ctx, sc, metav1.CreateOptions{})
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error ensuring secret for service account %s: %w", keyFromObject(sa), err)
 	}
-	if secret != nil {
-		return secret, nil
+
+	return secret, nil
+}
+
+// returns true if the secret has changed.
+func annotateSAWithSecret(ctx context.Context, sa *corev1.ServiceAccount, secret *corev1.Secret, saClient clientv1.ServiceAccountInterface, secretClient clientv1.SecretInterface) (*corev1.ServiceAccount, bool, error) {
+	if sa.Annotations[ServiceAccountSecretRefAnnotation] != "" {
+		// TODO: I guess we should check to see if it's the existing SA?
+		return sa, false, nil
 	}
+	if sa.Annotations == nil {
+		sa.Annotations = map[string]string{}
+	}
+	sa.Annotations[ServiceAccountSecretRefAnnotation] = secret.Namespace + "/" + secret.Name
 
-	sc := SecretTemplate(sa)
-	secret, err = secretClient.Create(ctx, sc, metav1.CreateOptions{})
+	updated, err := saClient.Update(ctx, sa, metav1.UpdateOptions{})
+	if err == nil {
+		return updated, false, nil
+	}
+	if !apierrors.IsConflict(err) {
+		logrus.Debugf("Successfully annotated ServiceAccount for %s", keyFromObject(sa))
+		return nil, false, err
+	}
+
+	// Rollback the optimistically created secret
+	logrus.Debugf("Rolling back ServiceAccount secret for %s", keyFromObject(secret))
+	if err := secretClient.Delete(ctx, secret.Name, metav1.DeleteOptions{}); err != nil {
+		return nil, false, fmt.Errorf("deleting optimistically locked secret for %s - %s: %w", keyFromObject(secret), keyFromObject(sa), err)
+	}
+	// Load the version that triggered the issue
+	logrus.Debugf("Reloading ServiceAccount %s", keyFromObject(sa))
+	updated, err = saClient.Get(ctx, sa.Name, metav1.GetOptions{})
 	if err != nil {
-		return nil, fmt.Errorf("error ensuring secret for service account [%s:%s]: %w", sa.Namespace, sa.Name, err)
+		return nil, false, fmt.Errorf("getting updated service account %s: %w", keyFromObject(sa), err)
 	}
 
-	return secret, nil
+	return updated, true, nil
+
+}
+
+func keyFromObject(obj namespacedObject) namespacedName {
```

Comment:
nitpick: I don't see other usage for `keyFromObject` other than logging and error messages that would justify using a struct. In fact, the implementation of the `String()` method is primarily focused on that use case (due to wrapping around square brackets).
I see other places where we could use a similar approach (e.g. [line 227](https://github.com/rancher/rancher/pull/47273/files#diff-8a853a257f4c9e8e5812b706a4f6a56eaa32ff67e57f9ac5ac4f78f1e1659f04R227)) directly use string concatenation:
```
sa.Annotations[ServiceAccountSecretRefAnnotation] = secret.Namespace + "/" + secret.Name
```
What about renaming this to explicitly mention logging and make it directly return a string? (beware also that `GetNamespace()` could return an empty string for cluster level objects, while still matching the interface)